src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp

```diff
@@ -571,68 +571,143 @@ void PDPServer::initializeParticipantProxyData(
+        EPROSIMA_LOG_ERROR(RTPS_PDP_SERVER, "Participant " << pdata.m_guid.guidPrefix
                                                            << " did not send information about builtin writers");
         return;
     }
 
     // only SERVER and CLIENT participants will be received. All builtin must be there
-    auxendp = endp & DISC_BUILTIN_ENDPOINT_PARTICIPANT_DETECTOR;
+    auxendp = endp & (DISC_BUILTIN_ENDPOINT_PARTICIPANT_DETECTOR | DISC_BUILTIN_ENDPOINT_PARTICIPANT_SECURE_DETECTOR);
     if (0 != auxendp)
     {
         auto temp_reader_data = get_temporary_reader_proxies_pool().get();
 
         temp_reader_data->clear();
         temp_reader_data->m_expectsInlineQos = false;
-        temp_reader_data->guid().guidPrefix = pdata->m_guid.guidPrefix;
-        temp_reader_data->guid().entityId = c_EntityId_SPDPReader;
-        temp_reader_data->set_remote_locators(pdata->metatraffic_locators, network, use_multicast_locators);
+        temp_reader_data->guid().guidPrefix = pdata.m_guid.guidPrefix;
+        temp_reader_data->guid().entityId = endpoints->reader.reader_->getGuid().entityId;
+        temp_reader_data->set_remote_locators(pdata.metatraffic_locators, network, use_multicast_locators);
         temp_reader_data->m_qos.m_reliability.kind = dds::RELIABLE_RELIABILITY_QOS;
         temp_reader_data->m_qos.m_durability.kind = dds::TRANSIENT_LOCAL_DURABILITY_QOS;
-        endpoints->writer.writer_->matched_reader_add(*temp_reader_data);
+#if HAVE_SECURITY
+        if (should_protect_discovery())
+        {
+            mp_RTPSParticipant->security_manager().discovered_builtin_reader(
+                endpoints->writer.writer_->getGuid(), pdata.m_guid,
+                *temp_reader_data, endpoints->writer.writer_->getAttributes().security_attributes());
+        }
+        else
+#endif // HAVE_SECURITY
+        {
+            endpoints->writer.writer_->matched_reader_add(*temp_reader_data);
+        }
     }
     else
     {
-        EPROSIMA_LOG_ERROR(RTPS_PDP_SERVER, "Participant " << pdata->m_guid.guidPrefix
+        EPROSIMA_LOG_ERROR(RTPS_PDP_SERVER, "Participant " << pdata.m_guid.guidPrefix
                                                            << " did not send information about builtin readers");
         return;
     }
+}
 
-    //Inform EDP of new RTPSParticipant data:
-    notifyAboveRemoteEndpoints(*pdata);
+void PDPServer::assignRemoteEndpoints(
+        ParticipantProxyData* pdata)
+{
+    EPROSIMA_LOG_INFO(RTPS_PDP_SERVER, "Assigning remote endpoint for RTPSParticipant: " << pdata->m_guid.guidPrefix);
+
+    match_reliable_pdp_endpoints(*pdata);
+
+#if HAVE_SECURITY
+    if (mp_RTPSParticipant->security_manager().discovered_participant(*pdata))
+#endif // HAVE_SECURITY
+    {
+        perform_builtin_endpoints_matching(*pdata);
+    }
 }
 
 void PDPServer::notifyAboveRemoteEndpoints(
         const ParticipantProxyData& pdata)
+{
+#if HAVE_SECURITY
+    match_reliable_pdp_endpoints(pdata);
```

Comment:
Methods `match_pdp_reader_nts_` and `match_pdp_writer_nts_` are meant for server to server matching. Method `match_reliable_pdp_endpoints` is meant for any participant with which the server should communicate. Different purpose, different methods.

I agree that it would be nice for the code of PDPClient and PDPServer to be as similar as possible, but I don't think it is possible in this case.